docs/design.md

```diff
@@ -0,0 +1,785 @@
+  name: unique-name-given-schema
+  storagePolicy: cleartext
+  labels:
+    genesis: enabled
+    master: enabled
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      required_key_a: required_label_a
+      required_key_b: required_label_b
+    actions:
+      - method: merge
+        path: .path.to.merge.into.parent
+      - method: delete
+        path: .path.to.delete
+  substitutions:
+    - dest:
+        path: .substitution.target
+      src:
+        schema: another-service/SourceType/v1
+        name: name-of-source-document
+        path: .source.path
+data:
+  path:
+    to:
+      merge:
+        into:
+          parent:
+            foo: bar
+          ignored:  # Will not be part of the resultant document after layering.
+            data: here
+  substitution:
+    target: null  # Paths do not need to exist to be specified as substitution destinations.
+...
+```
+
+##### schema: metadata/Control/v1
+
+This schema is the same as the `Document` schema, except it omits the
+`storagePolicy`, `layeringDefinition`, and `substitutions` keys, as these
+actions are not supported on `Control` documents.
+
+The complete list of valid `Control` document kinds is specified below along
+with descriptions of each document kind.
+
+##### schema: metadata/Tombstone/v1
+
+The only valid key in a `Tombstone` metadata section is `name`.  Additionally,
+the top-level `data` section should be omitted.
+
+### Layering
+
+Layering provides a restricted data inheritance model intended to help reduce
+duplication in configuration. Documents with different `schema`s are never
+layered together (see the Substitution section if you need to combine data
+from multiple types of documents).
+
+Layering is controlled in two places:
+
+1. The `LayeringPolicy` control document (described below), which defines the
+   valid layers and their order of precedence.
+2. In the `metadata.layeringDefinition` section of normal
+   (`metadata.schema=metadata/Document/v1`) documents.
+
+When rendering a particular document, you resolve the chain of parents upward
+through the layers, and begin working back down each layer rendering at each
+document in the chain.
+
+When rendering each layer, the list of `actions` will be applied in order.
+Supported actions are:
+
+* `merge` - a "deep" merge that layers new and modified data onto existing data
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
+After actions are applied for a given layer, substitutions are applied (see
+the Substitution section for details).
+
+Selection of document parents is controlled by the `parentSelector` field and
+works as follows. A given document, `C`, that specifies a `parentSelector`
+will have exactly one parent, `P`. Document `P` will be the highest
+precedence (i.e. part of the lowest layer) document that has the labels
```

Comment:
I had to read this a few times. The order of precedence is defined by data.layerOrder in the LayeringPolicy, so now this makes sense.